Feat/pod level opa (#96)
odahu/odahu-docs: source/comp_security.rst

======================
Security subsystem
======================

Security subsystem is distributed as a helm chart and relies on `OpenPolicyAgent`_ to make decisions about authorization
and `Istio`_ to enforce authorization for requests to the protected services.

> You can read concepts about ODAHU security in :ref:`an appropriate docs section<gen_security:Security>`

Reviewer: ```suggestion
You can read about ODAHU security concepts in :ref:`an appropriate docs section<gen_security:Security>`
```

Implementation details
----------------------

.. image:: img/security_comp_impl.png
   :scale: 80 %

Helm chart deploys

- Webhook server that injects `OpenPolicyAgent`_ sidecars into pods that labeled by ``odahu-flow-authorization=enabled``
- EnvoyFilter that configures Istio-proxy sidecars located in pods labeled by ``odahu-flow-authorization=enabled``
  to force authentication and authorization for all incoming http requests
- ConfigMap with common policy that included into all `OpenPolicyAgent`_ sidecars and implements masking of a sensitive data
- ConfigMap with default policy that included into `OpenPolicyAgent`_ sidecars when pod does not specify ConfigMap with its polices

> When webhook server deploys `OpenPolicyAgent`_ sidecar it attaches policies from ConfigMap. ConfigMap can be found by
> value of pod label ``opa-policy-config-map-name``. If this label is missed then default policy will be used.
> Default policy – reject all requests.

Reviewer 2: ```suggestion
When the webhook server deploys `OpenPolicyAgent`_ sidecar, it attaches policies from ConfigMap. ConfigMap can be found by
the value of the pod label ``opa-policy-config-map-name``. If this label is missed, then the default policy will be used.
Default policy – reject all requests.
```

Different ODAHU components such as :ref:`API <comp_api:API>` and :ref:`Feedback aggregator <comp_feedback:Feedback aggregator>`

> are distributed with a pre-defined set of OpenPolicyAgent_ policies. They create ConfigMap with their policies in the deployment process.

Reviewer: ```suggestion
are distributed with a pre-defined set of OpenPolicyAgent_ policies. They create ConfigMap with their policies during the deployment process.
```

> If user change ConfigMap with policies then appropriate pod must be restarted to refresh its policies.

Reviewer 2: ```suggestion
If you change ConfigMap with policies then the appropriate pod must be restarted to refresh its policies.
```
